### PR TITLE
Updated the SerializeContext `createuuid` options

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
### SerializeContextTools changes
Updated the createuuid options

Support has been added for a `--values-file` option which allows
specifying a text file which contains lines which should be converted to
Uuids using the SHA1 hash algorithm

With the support, was also added the ability to specify stdin as an
input file to the `--values-file` option using the `-` value.

The `--output-file` option for the createuuid, dumptypes, createtype
also supports using the value `-` to specify stdout as output file to
open

### SystemFile changes
Adding support for SystemFile API to support wraping Std handles

The `stdin`, `stdout` and `stderr` file handles are wrapped by static
functions on the SystemFile class that allows using the SystemFile class
to read from stdin or write to stdout/stderr using its API.

The Std handles are setup to not close on destruction of the SystemFile
class by default.

So it is safe to store these SystemFiles instances by value.

Finally a "skip close on destruction" flag has been added to the SystemFile open
mode flags to allow an open SystemFile to skip closing an open file
handle in the SystemFile destructor.

### Command Line parsing changes
Add support to CommandLine parsing to support the -- argument

The psuedo argument of `--` is used by the AZ::CommandLine parser to
force any arguments that come afterwards to be parsed as positional
arguments only

### Text Parser addition

Add TextParser class that can parse "line" delimited text entries

The TextParser supports AZ::IO::GenericStream classes and works
similarly to the AzFramework FileFunc ReadTextFile function.

However the TextParser supports additional configuration operations such
as changing the "text entry" delimiter, which is useful for NUL '\0'
delimited content as well as an option to skip over empty text entries.

The TextParser is being used by the SerializeContext `createuuid` command to read the list of line delimited string values from input files provided by the user using the code in [Dumper.cpp](https://github.com/o3de/o3de/blob/c4caf6717795c72595b196adcedc68e80abf1285/Code/Tools/SerializeContextTools/Dumper.cpp#L686-L718)

## How was this PR tested?

Added new UnitTest for the TextParser, CommandLine and SystemFile changes

Verified the SerializeContextTools `--values-file` option using a input file that progressive grew in content.
The first iteration of the `input.txt` file was empty.

input.txt - First iteration
---
```
```

input.txt - Second iteration
---
```
engine.json
```

input.txt - Third iteration(only difference is newline at the end of the file
---
```
engine.json

```

input.txt - Fourth iteration
---
```
engine.json
project.json
```

```PS
C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json
{B076CDDC-14DF-50F4-A5E9-7518ABB3E851} project.json

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt --output-file=-
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json
{B076CDDC-14DF-50F4-A5E9-7518ABB3E851} project.json

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=../input.txt --output-file=../output.txt

C:\code\o3de>build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-

C:\code\o3de>echo "engine.json" | build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-
{3506B64A-308F-5661-9BB5-E9E8EAC2F41E} "engine.json"

C:\code\o3de>echo engine.json | build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json

C:\code\o3de>printf "engine.json\nproject.json" | build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-

C:\code\o3de>printf "engine.json\nproject.json" | build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json
{B076CDDC-14DF-50F4-A5E9-7518ABB3E851} project.json

C:\code\o3de>printf "engine.json\nproject.json\n" | build\windows\bin\debug\SerializeContextTools.exe createuuid --values-file=-
{3B28A661-E723-5EBE-AB52-EC5829D88C31} engine.json
{B076CDDC-14DF-50F4-A5E9-7518ABB3E851} project.json
```

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp [hunk 1]
```diff
@@ -43,6 +47,19 @@ namespace AZ
...
+    {
+        //! Set to true if the psuedo argument '--' has been parsed
+        //! This indicates to the argument parser that the remaining
+        //! arguments should be treated as positional arguments only
+        bool m_parseRemainAsPositional{ false};
+        //! Stores the current option being parsed
+        //! This is used for maintaining state for space separated options
+        //! Ex. --foo bar
+        //! When the "bar" argument is parsed the current switch would be set to "foo"
+        AZStd::string m_currentSwitch;
```
Comment (forkercat) [inline Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp:60]: question: Thinking can we change it to `m_currentOption`. Is 'switch' still needed to indicate something more than 'option' does?

**Review by forkercat — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Settings/CommandLine.cpp:60]: I'll change it to the option.

File: Code/Framework/AzCore/AzCore/Settings/CommandLine.h [hunk 2]
```diff
@@ -19,14 +19,28 @@ namespace AZ
...
+    * https://docs.python.org/3/library/argparse.html#action
+    * For example the command line of
+    * `app.exe --verbose <PositionalArg>` would parse <PositionalArg> as the value of the "verbose" flag
+    * The workaround is to either
+    * 1. specify the flag options at the end of the command line
+    * `app.exe <PositionalArg> --verbose`
+    * 2. specify second option right after the previous flag option with and add a value to it
+    * `app.exe --verbose --count 2 <PositionalArg>`
+    * 3. specify a value for the flag option
+    * `app.exe --verbose 1 <PositionalArg>`  or `app.exe --verbose=1 <PositionalArg>`
```
[the remark this answers is not in the document]
Comment (forkercat) [reply Code/Framework/AzCore/AzCore/Settings/CommandLine.h]: Got it. Thanks for the explanation.

**Review by lawsonamzn — APPROVED** [review]
excellent.  Will there be any quick blurb about how to use the new text parser?

**Comment by lumberyard-employee-dm** [comment]
> excellent. Will there be any quick blurb about how to use the new text parser?

Yes I can add one to the o3de.org documentation.
I created a ticket with the content to add to example for the TextParser: https://github.com/o3de/o3de.org/issues/2356

**Pull request merged**